pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -111,17 +130,97 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, t
 	return controller.Result{}, nil
 }
 
-// ReconcileControlPlane applies the control plane CAPI objects to the cluster.
-func (r *Reconciler) ReconcileControlPlane(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+// GenerateSpec generates Tinkerbell control plane and workers spec.
+func (r *Reconciler) GenerateSpec(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
 	spec := tinkerbellScope.ClusterSpec
-	log = log.WithValues("phase", "reconcileControlPlane")
-	log.Info("Applying control plane CAPI objects")
+	log = log.WithValues("phase", "GenerateSpec")
+	log.Info("Generating tinkerbell control plane and workers spec")
+
 	cp, err := tinkerbell.ControlPlaneSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
 	if err != nil {
 		return controller.Result{}, err
 	}
+	tinkerbellScope.ControlPlane = cp
+
+	w, err := tinkerbell.WorkersSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating workers spec")
+	}
+	tinkerbellScope.Workers = w
 
-	return clusters.ReconcileControlPlane(ctx, r.client, toClientControlPlane(cp))
+	return controller.Result{}, nil
+}
+
+// DetectOperation detects change type.
+func (r *Reconciler) DetectOperation(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+	log = log.WithValues("phase", "DetectOperation")
+	log.Info("Detecting operation type")
+
+	currentKCP, err := controller.GetKubeadmControlPlane(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+	if currentKCP == nil {
+		tinkerbellScope.ClusterChange = ReconcileNewCluster
+		return controller.Result{}, nil
+	}
+
+	tinkCurrentKCP := tinkerbell.KubeadmControlPlane{KubeadmControlPlane: currentKCP}
+	tinkDesiredKCP := tinkerbell.KubeadmControlPlane{KubeadmControlPlane: tinkerbellScope.ControlPlane.KubeadmControlPlane}
+	wantVersionChange := tinkerbell.HasKubernetesVersionChange(&tinkCurrentKCP, &tinkDesiredKCP)
+	cpWantScaleChange := tinkerbell.ReplicasDiff(&tinkCurrentKCP, &tinkDesiredKCP)
+	workerWantScaleChange := r.workerReplicasDiff(ctx, tinkerbellScope)
```

Comment:
Looking at this, IMO the `ClusterChange` adds unnecessary complexity.

The only thing that is caching is the result of a couple if else's that compare strings and int's that have been already fetched.
Unless we can prove that's having a performance impact, I would vote to remove it and just make the calculation twice. Specially now that is encapsulated in the tinkerbell package.

Reply:
In order to get the if else result, we also get the current capi cp and md from client. So basically trade off between get those capi objects multiple times or cache the result

Reply:
We're caching either way. If we need to cache the current state for later then we could have a `scope.GetOperation()` and remove this phase.

Reply:
I think running a something like GetOperation() works, but we would still need to retrieve the current state CAPI objects to cache them for comparison. Does it make sense to keep the phase around to retrieve the current state CAPI? We still need the current state CAPI for the `AssertionsForScaleUpDown` validation.

Reply:
In case it helps with the conversation, the client we use is a cached client. So, without going too much into how its cache invalidation works (which I'm not an expert in either), it's safe to assume the same object won't be read twice in the same reconciliation loop the majority of the time.

That said, if you still want to "cache" something in the scope to avoid going through the client's cache (which might or might not be more inefficient), I think caching the objects themselves (current CAPI state) is a better idea than caching the "change" type. A couple reasons:
* It's simpler.
* It's more reusable. The current cluster change only allows for one change at the same time. By introducing some assumptions in the abstraction (like you can't have scale up/down at the same time as a kubernetes upgrade), it becomes a bit inflexible.
* It doesn't create a new abstraction.

I think this might be one of those times where we tried to optimized too early. If the only gain between the set of solutions we are considering is performance, my vote is to implement whatever is simpler and then measure performance once we have a working controller.